File: ChelaVm/src/Interface.cpp
#include "ChelaVm/Interface.hpp"

namespace ChelaVm
{
    Interface::Interface(Module *module)
        : Structure(module)
    {
    }
    
    Interface::~Interface()
    {
    }
        
    bool Interface::IsStructure() const
    {
        return false;
    }
    
    bool Interface::IsClass() const
    {
        return false;
    }

    bool Interface::IsInterface() const
    {
        return true;
    }

    bool Interface::IsPassedByReference() const
    {
        return true;
    }

    Interface *Interface::PreloadMember(Module *module, ModuleReader &reader, const MemberHeader &header)
    {
        // Create the result.
        Interface *res = new Interface(module);

        // Store the name and the member flags.
        res->name = module->GetString(header.memberName);
        res->flags = (MemberFlags)header.memberFlags;

        // Skip the member data.
        reader.Skip(header.memberSize);
        return res;
    }

    Structure *Interface::CreateBuilding(Module *module) const
    {
        return new Interface(module);
    }
};

File: ChelaCompiler/Module/Interface.cs
namespace Chela.Compiler.Module
{
    public class Interface: Structure
    {
        protected Interface(ChelaModule module)
            : base(module)
        {
        }

        public Interface (string name, MemberFlags flags, Scope parentScope)
            : base(name, flags, parentScope)
        {
        }

        public override bool IsInterface ()
        {
            return true;
        }

        public override bool IsStructure ()
        {
            return false;
        }

        public override bool IsClass ()
        {
            return false;
        }

        public override bool IsPassedByReference ()
        {
            return true;
        }

        internal new static void PreloadMember(ChelaModule module, ModuleReader reader, MemberHeader header)
        {
            // Create the temporal interface and register it.
            Interface iface = new Interface(module);
            module.RegisterMember(iface);

            // Read the name.
            iface.name = module.GetString(header.memberName);
            iface.flags = (MemberFlags)header.memberFlags;

            // Skip the class elements.
            reader.Skip(header.memberSize);
        }
    }
}

